internal/controllers/gateway/gateway_controller.go

```diff
@@ -0,0 +1,581 @@
+	return svc, r.Client.Get(ctx, r.publishServiceRef, svc)
+}
+
+// determineL4ListenersFromService generates L4 addresses and listeners for a
+// unmanaged Gateway provided the service to reference from.
+func (r *GatewayReconciler) determineL4ListenersFromService(
+	svc *corev1.Service,
+) (
+	[]gatewayv1alpha2.GatewayAddress,
+	[]gatewayv1alpha2.Listener,
+	error,
+) {
+	// if there are no clusterIPs available yet then this service
+	// is still being provisioned so we will need to wait.
+	if len(svc.Spec.ClusterIPs) < 1 {
+		return nil, nil, fmt.Errorf("gateway service %s/%s is not yet ready (no cluster IPs provisioned)", svc.Namespace, svc.Name)
+	}
+
+	// take var copies of the address types so we can take pointers to them
+	gatewayIPAddrType := gatewayv1alpha2.IPAddressType
+	gatewayHostAddrType := gatewayv1alpha2.HostnameAddressType
+
+	// for all service types we're going to capture the ClusterIP
+	addresses := make([]gatewayv1alpha2.GatewayAddress, 0, len(svc.Spec.ClusterIPs))
+	listeners := make([]gatewayv1alpha2.Listener, 0, len(svc.Spec.Ports))
+	for _, clusterIP := range svc.Spec.ClusterIPs {
+		for _, port := range svc.Spec.Ports {
+			addresses = append(addresses, gatewayv1alpha2.GatewayAddress{
+				Type:  &gatewayIPAddrType,
+				Value: clusterIP,
+			})
+			listeners = append(listeners, gatewayv1alpha2.Listener{
+				Name:          gatewayv1alpha2.SectionName(port.Name),
+				Protocol:      gatewayv1alpha2.ProtocolType(port.Protocol),
+				Port:          gatewayv1alpha2.PortNumber(port.Port),
+				AllowedRoutes: &r.allowedRoutes,
+			})
+		}
+	}
+
+	// for LoadBalancer service types we'll also capture the LB IP or Host
+	if svc.Spec.Type == corev1.ServiceTypeLoadBalancer {
+		// if the loadbalancer IPs/Hosts haven't been provisioned yet
+		// the service is not ready and we'll need to wait.
+		if len(svc.Status.LoadBalancer.Ingress) < 1 {
+			return nil, nil, fmt.Errorf("gateway service %s/%s is not yet ready (no LoadBalancer IP/Host provisioned)", svc.Namespace, svc.Name)
+		}
+
+		// otherwise gather any IPs or Hosts provisioned for the LoadBalancer
+		// and record them as Gateway Addresses.
+		for _, ingress := range svc.Status.LoadBalancer.Ingress {
+			if ingress.IP != "" {
+				addresses = append(addresses, gatewayv1alpha2.GatewayAddress{
+					Type:  &gatewayIPAddrType,
+					Value: ingress.IP,
+				})
+			}
+			if ingress.Hostname != "" {
+				addresses = append(addresses, gatewayv1alpha2.GatewayAddress{
+					Type:  &gatewayHostAddrType,
+					Value: ingress.Hostname,
+				})
+			}
+		}
+	}
+
+	return addresses, listeners, nil
+}
+
+// determineL6and7ListenersFromDataPlane takes a list of L4 Gateway listeners and references
+// them against the data-plane to determine any L6 (TLS) or L7 (HTTP, e.t.c.) support configured for them.
+func (r *GatewayReconciler) determineL6and7ListenersFromDataPlane(ctx context.Context, svc *corev1.Service, listeners []gatewayv1alpha2.Listener) ([]gatewayv1alpha2.Listener, error) {
```

Comment:
I don't see why TLS would be L6 and not L5.

In general the OSI layer metaphore doesn't stick with me, but as long as it's explained clearly that's fine.

This is not a request for change, just inspiration: my take on the nomenclature here would be something like replacing:
- "L4" and "L6" with "HostPort" and "Stream"
- "L7" with "RequestProxy" or something similar.

Reply:
This is I think a source of debate to this day, so rather than worry about it I've changed the language to not reference the model 17392644e3d9ef49998badb47cc385f56da92370